### PR TITLE
Snapshot Conflicts 

### LINKED ISSUE #22: Show conflicts arising with scheduled snapshots
When making a change in the Customizer or making a new snapshot, if there is currently one or more scheduled snapshots that touch on a setting that is dirty in the current state, there should be a warning to note of a potential conflict.

See also previewing future scheduled snapshots: #15 & #21

### PR DESCRIPTION
**Pull request by PatelUtkarsh**
Fixes #22

### CODE REVIEW

File: php/class-post-type.php
Comment (valendesigns) [inline php/class-post-type.php]: This conditional would probably make more sense like so:

``` PHP
if ( ! $post || ! ( $post instanceof \WP_Post ) ) {
```

File: php/class-post-type.php
Comment (valendesigns) [inline php/class-post-type.php]: Not sure the `1 = 2` is needed here?

File: css/customize-snapshots-admin.css
Comment (valendesigns) [inline css/customize-snapshots-admin.css]: Add a space before curly braces.

File: php/class-post-type.php
Comment (valendesigns) [inline php/class-post-type.php]: Change the link to:

``` PHP
$title_text = sprintf( esc_html__( '%s has potential Snapshot conflicts (click to expand)', 'customize-snapshots' ), $setting_id );
echo '<a href="#TB_inline?width=600&height=550&inlineId=snapshot-' . esc_attr( $setting_id_key ) . '" class="dashicons dashicons-warning thickbox snapshot-thickbox" title="' . esc_attr( $title_text ) . '"></a>';
```

File: php/class-post-type.php
Comment (valendesigns) [inline php/class-post-type.php]: The are too many tabs

File: php/class-post-type.php
Comment (PatelUtkarsh) [inline php/class-post-type.php]: By adding that we can directly append OR in `$query .= $wpdb->prepare( 'OR post_content LIKE %s', '%' . $wpdb->esc_like( wp_json_encode( $setting_id ) ) . '%' );`
However i have updated code using `implode` and removed `1 = 2`

**Comment by westonruter** [comment]
Will review once #68 is merged.

File: php/class-customize-snapshot-manager.php
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: `sprintf` isn't needed here the way this string is generated.

File: php/class-customize-snapshot-manager.php
Comment (valendesigns) [inline php/class-customize-snapshot-manager.php]: Why do you need to add `style="display:none;"`?

Comment (PatelUtkarsh) [inline php/class-customize-snapshot-manager.php]: `thickbox` is initially hidden on click of conflict icon it should open Thickbox. I will move this css to css file.

**Comment by westonruter** [comment]
As for embedding the conflict button when a control is embedded, you could take this approach:

``` js
component.addConflictButton = function addConflictButton( control ) {
    control.deferred.embedded.done( function() {
        var onFirstChange, hasDirty;

        if ( ! control.setting ) {
            return;
        }

        onFirstChange = function() {
            _.each( control.settings, function( setting ) {
                setting.unbind( onFirstChange );
            } );

            // @todo Inject notification button template (even hidden) into control.container somewhere.
            // @todo Do Ajax request to obtain potential conflicts for control.settings
            // @todo The call to do the Ajax request for settings should be batched, so that when loading a dirty snapshot, all of the settings are requested in one Ajax request. See wp.customize.selectiveRefresh.requestPartial for an example of this.
        };

        hasDirty = _.find( control.settings, function( setting ) {
            return setting._dirty;
        } );
        if ( hasDirty ) {
            onFirstChange();
        } else {
            _.each( control.settings, function( setting ) {
                setting.bind( onFirstChange );
            } );
        }
    } )
};

wp.customize.control.bind( 'add', component.addConflictButton );
wp.customize.control.each( component.addConflictButton );
```

**Comment by westonruter** [comment]
@lgedeon @PatelUtkarsh just to clarify, even for myself. What Luke is working on in https://github.com/xwp/wp-customize-concurrency/pull/4 is closely related but distinct from what is being worked on here.

This PR here (#71) is about there is one for showing ​_potential_​ conflicts with ​_other_​ snapshots.
Luke's https://github.com/xwp/wp-customize-concurrency/pull/4 is about _actual_ conflicts _published_ changes. Once that in place, then he'll be able to move on to showing _actual_ conflicts with _saved_ changes to the _same_ snapshot, when someone else may be editing the same snapshot in another session. That would be tackled in #60.

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: Instead of “edit” how about “inspect”? Also, the parents should be _inside_ the translated string, and best to use `esc_html_e()` here.

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: There is a danger here for adding a button to the title because a control may actually not have a title. However, all controls will have a… notification area. I just realized an alternative to adding a custom button like this it could be better to make use of the notification area. A notification of type `warning` could be added as depicted here: https://github.com/xwp/wp-customize-posts/issues/217#issuecomment-239746179

File: php/class-post-type.php
Comment (PatelUtkarsh) [inline php/class-post-type.php]: @westonruter @valendesigns which post_status should we check for conflict? Currently, I am checking all 3 `'draft', 'pending', 'future'` I think we should remove `pending` but should we check `draft` or not?

Comment (westonruter) [reply php/class-post-type.php]: @PatelUtkarsh I think only `future` or `pending`. If you check `draft` then the conflict should be “informative” rather than a warning.

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: Use `api` instead of `wp.customize` in this case and the other cases in this PR.

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Missing `esc_html()` here.

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Best to use `esc_html_e()` here instead of `_e()`. Also, I think the parens should be included in the translated string.

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Should the `customize_snapshot_value_preview` filter apply on this as well?

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: PHP keys should be `snake_case` I think. So `id` instead of `ID` and `edit_link` instead of `editLink`.

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: I think `$return` should be renamed to something more descriptive, like `$conflicted_settings`.

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Maybe `get_conflicted_settings()` would be a better method name?

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Either the post type for `$post` should be checked to be `customize_snapshot` or the a `Customize_Snapshot` could be passed as the argument instead.

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Is this not too late? If it is after `wp_head` does this not mean enqueued styles miss their chance to get printed?

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: Should do `wp_unslash()` on `$_POST['setting_ids']`.

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: Why not just use `data.setting_id` as-is?

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: This string: `#TB_inline?width=600&height=550&inlineId=snapshot-conflicts-{{id}}` should be entity-escaped (e.g. `&amp;`)

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: Is `control.notificationsTemplate = component.conflict.notificationTemplate` doing anything yet?

Comment (PatelUtkarsh) [reply js/customize-snapshots.js]: Yes we are overriding `control.notificationsTemplate` so it allows markup as a notification message.

File: php/class-post-type.php
Comment (PatelUtkarsh) [reply php/class-post-type.php]: Yes moving filter inside `get_printable_setting_value`
On a second thought, we won't have filter param values inside that.

File: php/class-customize-snapshot-manager.php
Comment (PatelUtkarsh) [reply php/class-customize-snapshot-manager.php]: It will go to html tag's `id` attribute and it can not have `[]` which setting id will have.

**Comment by mohdsayed** [comment]
tested, working as expected 👍

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [reply php/class-customize-snapshot-manager.php]: Why can't an HTML `id` attribute have brackets? There are no restrictions on what an `id` can have, except it cannot have spaces. You may be recalling HTML4 or XML which restricted the characters which can be in an ID. But browsers were always forgiving here, and HTML5 says the `id` attribute can contain anything. See https://www.w3.org/TR/html5/dom.html#the-id-attribute

image

**Comment by westonruter** [comment]
@PatelUtkarsh there are conflicts now in this PR. Would you please resolve?

File: php/class-customize-snapshot-manager.php
Comment (PatelUtkarsh) [reply php/class-customize-snapshot-manager.php]: Thank you for pointing that out I was in impression that it can not have `[]` i will update the code and fix this.

**Comment by valendesigns** [comment]
@PatelUtkarsh Is this PR done, or are you waiting for feedback?

**Comment by PatelUtkarsh** [comment]
@valendesigns This PR is ready, Waiting for review.

File: js/customize-snapshots.js
Comment (PatelUtkarsh) [inline js/customize-snapshots.js]: @westonruter Sometimes `setting.findControls();` throws an error. 

`customize-post-section.js?ver=0.7.1-alpha:327Uncaught TypeError: Cannot read property 'call' of undefined` 
Last call trace https://github.com/xwp/wordpress-develop/blob/4.5/src/wp-includes/js/customize-base.js#L565-L567

**Review by PatelUtkarsh — COMMENTED** [review]
Needs fix before merge putting PR in WIP again

File: js/customize-snapshots.js
Comment (PatelUtkarsh) [reply js/customize-snapshots.js]: Okay, that was because I was running this on WP 4.5.x. 
We need to safely fall when the version is below 4.6.

**Comment by valendesigns** [comment]
@PatelUtkarsh Please resolve conflicts again when you get a chance. I'll test and review after. Also update the `dev-lib` which fixes the coverage reports not working since March.

**Comment by westonruter** [comment]
I haven't had the chance to review this yet, but I did want to bring up some related work that @aaronjorbin brought to my attention. In particular, the whole “[post forking](https://github.com/post-forking/post-forking)” project is very similar conceptually to changeset forking and merging. 

While the post forking project has largely ceased, he noted the merging library worked really well:  https://github.com/post-forking/post-forking/blob/master/includes/diff3.php

What's more is that I notice this makes use of CodeMirror, and CodeMirror is now in core. That means that Snapshots could provide a changeset merging interface now like this: https://codemirror.net/demo/merge.html

**Comment by PatelUtkarsh** [comment]
@westonruter One issue this would have is, Snapshot allows merging multiple posts so for example if we are merging 10 posts and if 1 particular setting have conflicts with 4 post then showing in this way would be difficult. 
Let me know your thoughts.

**Comment by westonruter** [comment]
@PatelUtkarsh good point! Using CodeMirror in merging would only be suitable when merging 2 changesets, then. It's probably not necessary since you've developed a separate UI for it which is probably better and more tailored for this application anyway.

**Comment by PatelUtkarsh** [comment]
Todo:

- [x] Saved changeset setting should show conflicts when loaded. Check for `setting.findControls()` code.
- [x] Thickbox z-index issue. 
- [x] Current changeset should be skipped from the comparison from MySQL query as it's being added via JS.
- [x] Above Should Fix unit-test else check unit-test
- [x] Admin CSS updates and test.
- [x] Test.

cc: @sayedtaqui

**Comment by PatelUtkarsh** [comment]
I think this is ready for review. @sayedtaqui

**Comment by westonruter** [comment]
I plan to finally review this after WCUS!

File: php/class-customize-snapshot-manager-compat.php
Comment (westonruter) [inline php/class-customize-snapshot-manager-compat.php]: Should this not get exported via the `customize_refresh_nonces` filter?

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: This will be problematic to translate.

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: Indentation is off here a bit.

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [inline php/class-customize-snapshot-manager.php]: Why the `temp` variable here and above? You should be able to do just `{{ JSON.stringify( data.value, null, 4 ) }}`

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: jsdoc should have the description starting with capital. A hyphen can separate the param name from description:

```
@param {jQuery} selector - Control container or conflict markup container.
```

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: This is problematic to translate

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: A backslash should be used instead of a forward slash. Also, the param descriptions should be aligned vertically.

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Align descriptions and use capital to start description. For example:

```
 * @param \WP_Post $post     Post to compare conflict values.
 * @param array    $settings Setting to search for optional.
```

File: php/class-post-type.php
Comment (westonruter) [inline php/class-post-type.php]: Is this todo done?

Comment (PatelUtkarsh) [reply php/class-post-type.php]: Do we want to keep draft or pending and future are fine? This will filter post which we will be searching conflicted settings in. @westonruter

File: php/class-post-type.php
Comment (mohdsayed) [reply php/class-post-type.php]: @westonruter Did you mean, use `_x()` because the same string has been used before? https://github.com/xwp/wp-customize-snapshots/pull/71#discussion_r155712820
or using `(` `)` is problematic?

Comment (westonruter) [reply php/class-post-type.php]: What if a language is Object-Verb-Subject order? The placeholder should be inside the translation string.

File: php/class-post-type.php
Comment (westonruter) [reply php/class-post-type.php]: Limiting to `pending` and `future` make sense to me. I could also see `draft` being included.  Maybe include `draft` for now and then we can remove it if it gets too noisey?

File: php/class-customize-snapshot-manager.php
Comment (westonruter) [reply php/class-customize-snapshot-manager.php]: What if a language is Object-Verb-Subject order? The placeholder should be inside the translation string.

File: php/class-post-type.php
Comment (mohdsayed) [reply php/class-post-type.php]: Great point @westonruter  💯

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: I see that `addConflictButton` is getting re-called every time a `saved` happens, and here it is re-binding the `updateCurrentValue` handler to the setting. So that means this `updateCurrentValue` is going to get added again to the setting every time the changeset is saved. Shouldn't this be prevented?

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: This should probably check if `setting.extended( api.Setting )`

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: This should make sure that `setting.extended( api.Setting )`

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: This anonymous function is being bound here, and then there seems to be an attempt to unbind elsewhere with `setting.unbind( snapshot.handleConflictRequestOnFirstChange );` but it won't work because it's not the same function. I think a different approach to doing once is needed here, for example:

```diff
diff --git a/js/customize-snapshots.js b/js/customize-snapshots.js
index cf94a3b..0082b5b 100644
--- a/js/customize-snapshots.js
+++ b/js/customize-snapshots.js
@@ -347,7 +347,6 @@
 			var snapshot = this;
 
 			_.each( control.settings, function( setting ) {
-				setting.unbind( snapshot.handleConflictRequestOnFirstChange );
 				snapshot.handleConflictRequest( setting, control );
 			} );
 		},
@@ -386,10 +385,17 @@
 					bindFirstChange = true;
 				}
 				_.each( control.settings, function( setting ) {
+					var changeOnce;
 					if ( bindFirstChange ) {
-						setting.bind( function() {
+						changeOnce = function () {
 							snapshot.handleConflictRequestOnFirstChange( control );
-						} );
+							setting.unbind( changeOnce );
+						};
+						setting.bind( changeOnce );
 					}
 					setting.bind( updateCurrentValue );
 				} );
```

Comment (westonruter) [reply js/customize-snapshots.js]: But this can't get reset in `resetConflicts` since the function reference will not be available.

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: Per below, I don't think this is going to work. There are no calls to `setting.bind( snapshot.handleConflictRequestOnFirstChange )` to unbind.

File: js/customize-snapshots.js
Comment (westonruter) [inline js/customize-snapshots.js]: Per below, I don't think this is going to work. There are no calls to `setting.bind( snapshot.handleConflictRequestOnFirstChange )` to unbind.

**Review by westonruter — CHANGES_REQUESTED** [review]
I think these are the last changes needed.

**Review by westonruter — APPROVED** [review]

**Pull request merged**
